File: src/lib/ics.ts
function pad(n: number, width = 2): string {
  return String(n).padStart(width, '0')
}

function formatUtcBasic(d: Date): string {
  return (
    `${d.getUTCFullYear()}${pad(d.getUTCMonth() + 1)}${pad(d.getUTCDate())}` +
    `T${pad(d.getUTCHours())}${pad(d.getUTCMinutes())}${pad(d.getUTCSeconds())}Z`
  )
}

function formatDateBasic(dateStr: string): string {
  return dateStr.replace(/-/g, '')
}

function nextDayBasic(dateStr: string): string {
  const [y, m, d] = dateStr.split('-').map(Number)
  const next = new Date(Date.UTC(y, m - 1, d + 1))
  return `${next.getUTCFullYear()}${pad(next.getUTCMonth() + 1)}${pad(next.getUTCDate())}`
}

function escapeText(value: string): string {
  return value
    .replace(/\\/g, '\\\\')
    .replace(/;/g, '\\;')
    .replace(/,/g, '\\,')
    .replace(/\n/g, '\\n')
}

function byteLength(str: string): number {
  return new TextEncoder().encode(str).length
}

function foldLine(line: string): string {
  const LIMIT = 75
  if (byteLength(line) <= LIMIT) return line

  const chunks: string[] = []
  let current = ''
  let currentBytes = 0

  for (const char of line) {
    const charBytes = byteLength(char)
    const isContinuation = chunks.length > 0
    const limit = isContinuation ? LIMIT - 1 : LIMIT
    if (currentBytes + charBytes > limit) {
      chunks.push(current)
      current = ''
      currentBytes = 0
    }
    current += char
    currentBytes += charBytes
  }
  if (current) chunks.push(current)

  return chunks.join('\r\n ')
}

export type IcsEvent = {
  uid: string
  title: string
  description?: string | null
  location?: string | null
  url: string
  start: { date: string } | { dateTime: string; endDateTime?: string | null }
}

function buildVevent(e: IcsEvent, now: Date): string[] {
  let dtstart: string
  let dtend: string
  if ('date' in e.start) {
    dtstart = `DTSTART;VALUE=DATE:${formatDateBasic(e.start.date)}`
    dtend = `DTEND;VALUE=DATE:${nextDayBasic(e.start.date)}`
  } else {
    const start = new Date(e.start.dateTime)
    const end = e.start.endDateTime
      ? new Date(e.start.endDateTime)
      : new Date(start.getTime() + 60 * 60 * 1000)
    dtstart = `DTSTART:${formatUtcBasic(start)}`
    dtend = `DTEND:${formatUtcBasic(end)}`
  }

  const lines: string[] = [
    'BEGIN:VEVENT',
    `UID:${escapeText(e.uid)}`,
    `DTSTAMP:${formatUtcBasic(now)}`,
    dtstart,
    dtend,
    `SUMMARY:${escapeText(e.title)}`,
  ]

  if (e.description) lines.push(`DESCRIPTION:${escapeText(e.description)}`)
  if (e.location) lines.push(`LOCATION:${escapeText(e.location)}`)
  lines.push(`URL:${escapeText(e.url)}`)
  lines.push('END:VEVENT')

  return lines
}

export function buildIcs(e: IcsEvent & { now?: Date }): string {
  const now = e.now ?? new Date()
  const lines: string[] = [
    'BEGIN:VCALENDAR',
    'VERSION:2.0',
    'PRODID:-//samla//EN',
    'CALSCALE:GREGORIAN',
    'METHOD:PUBLISH',
    ...buildVevent(e, now),
    'END:VCALENDAR',
  ]

  return lines.map(foldLine).join('\r\n') + '\r\n'
}

/**
 * Same wire format as `buildIcs`, but with one VEVENT per input event inside a single VCALENDAR —
 * used for the sign-up claim confirmation email, where a participant may hold several slots at
 * once and a mail client should be able to add all of them from one attachment.
 */
export function buildIcsMulti(events: IcsEvent[], opts: { now?: Date } = {}): string {
  const now = opts.now ?? new Date()
  const lines: string[] = [
    'BEGIN:VCALENDAR',
    'VERSION:2.0',
    'PRODID:-//samla//EN',
    'CALSCALE:GREGORIAN',
    'METHOD:PUBLISH',
    ...events.flatMap((e) => buildVevent(e, now)),
    'END:VCALENDAR',
  ]

  return lines.map(foldLine).join('\r\n') + '\r\n'
}

File: internal/ics/ics.go
// Package ics is the shared iCalendar (RFC 5545) core internal/polls/ics.go and
// internal/bookings/ics.go each used to carry their own separate copy of: TEXT-value escaping,
// 75-octet line folding, the UTC/date basic date-time formats, and the CRLF-joined
// VCALENDAR...VEVENT...VCALENDAR wrapper. Both callers ported the same handful of RFC 5545 rules
// from the same TS source (src/lib/ics.ts's escapeText/foldLine/buildIcs) independently — one
// small package now backs both, so a rule only ever needs fixing in one place.
//
// What stays OUT of this package, deliberately: building a single event's own BEGIN:VEVENT..
// END:VEVENT line list is still each caller's own job (polls' buildVeventLines maps a finalized
// poll option — which can be a bare all-day date OR a timed datetime, per icsStart's own union —
// booking's BuildBookingICS maps a booking, which is always timed, never all-day). Neither
// caller's own event shape belongs in a package meant to serve both.
package ics

import (
	"strings"
	"time"
)

// LineLimit is RFC 5545's 75-octet content-line cap (CRLF excluded) that FoldLine enforces.
const LineLimit = 75

// textEscaper escapes the four characters iCalendar TEXT values must escape, backslash first (so
// later escapes' own backslashes aren't re-escaped), then semicolon, comma, and newline — ports
// ics.ts's escapeText exactly. strings.Replacer applies all pairs in a single left-to-right pass
// over the *original* string, which — since none of these four patterns can appear inside
// another's replacement — is equivalent to the TS chain of four sequential .replace() calls.
var textEscaper = strings.NewReplacer(
	`\`, `\\`,
	`;`, `\;`,
	`,`, `\,`,
	"\n", `\n`,
)

// EscapeText escapes s for use as an iCalendar TEXT value (SUMMARY, DESCRIPTION, LOCATION, a UID,
// ...).
func EscapeText(s string) string {
	return textEscaper.Replace(s)
}

// FoldLine ports ics.ts's foldLine: a content line longer than LineLimit octets is split into
// that first line plus continuation lines, each joined by "\r\n " (a CRLF followed by a single
// leading space, which iCalendar readers know to strip and rejoin) — never splitting a multi-byte
// UTF-8 rune across two lines. The first continuation's own octet budget is one less than every
// other line's, since the leading space that will precede it on rejoining still counts against
// the 75-octet cap for that physical line.
func FoldLine(line string) string {
	if len(line) <= LineLimit {
		return line
	}

	var chunks []string
	var current strings.Builder
	currentBytes := 0

	for _, r := range line {
		char := string(r)
		charBytes := len(char)
		limit := LineLimit
		if len(chunks) > 0 {
			limit = LineLimit - 1
		}
		if currentBytes+charBytes > limit {
			chunks = append(chunks, current.String())
			current.Reset()
			currentBytes = 0
		}
		current.WriteString(char)
		currentBytes += charBytes
	}
	if current.Len() > 0 {
		chunks = append(chunks, current.String())
	}

	return strings.Join(chunks, "\r\n ")
}

// FormatUTCBasic renders t in UTC as iCalendar's basic date-time form, e.g. "20260901T163000Z" —
// for a timed DTSTART/DTEND/DTSTAMP.
func FormatUTCBasic(t time.Time) string {
	return t.UTC().Format("20060102T150405Z")
}

// FormatDateBasic renders t's (UTC) calendar date in iCalendar's basic date form, e.g. "20260901"
// — for an all-day event's own DTSTART;VALUE=DATE.
func FormatDateBasic(t time.Time) string {
	return t.UTC().Format("20060102")
}

// NextDayBasic renders the day after t's (UTC) calendar date, in the same basic date form — an
// all-day event's DTEND is exclusive per RFC 5545, so a one-day event's own DTEND is the
// following date.
func NextDayBasic(t time.Time) string {
	return t.UTC().AddDate(0, 0, 1).Format("20060102")
}

// BuildCalendar wraps veventLines (one caller-built event's own, unfolded BEGIN:VEVENT..
// END:VEVENT lines — buildVeventLines in polls/ics.go, or BuildBookingICS in bookings/ics.go) in
// one VCALENDAR, folding every line (header, VEVENT, and footer alike) to LineLimit and joining
// with CRLF, with a trailing CRLF — ports ics.ts's buildIcs.
func BuildCalendar(veventLines []string) []byte {
	lines := []string{
		"BEGIN:VCALENDAR",
		"VERSION:2.0",
		"PRODID:-//whenweall//EN",
		"CALSCALE:GREGORIAN",
		"METHOD:PUBLISH",
	}
	lines = append(lines, veventLines...)
	lines = append(lines, "END:VCALENDAR")

	folded := make([]string, len(lines))
	for i, l := range lines {
		folded[i] = FoldLine(l)
	}
	return []byte(strings.Join(folded, "\r\n") + "\r\n")
}
